message/legacy/encode.go

```diff
@@ -6,182 +6,214 @@ package legacy
 
 import (
 	"bytes"
-	"encoding/json"
+	"errors"
 	"fmt"
-	"io"
-	"strconv"
+	jsoniter "github.com/json-iterator/go"
 	"strings"
 )
 
-func (pp *prettyPrinter) formatArray(depth int) error {
+var replacer = strings.NewReplacer("\\", `\\`, "\t", `\t`, "\n", `\n`, "\r", `\r`, `"`, `\"`)
+
+var iteratorConfig = jsoniter.Config{
+	// re float encoding: https://spec.scuttlebutt.nz/datamodel.html#signing-encoding-floats
+	// not particular excited to implement all of the above
+	// this keeps the original value as a string
+	UseNumber: true,
+}.Froze()
 
-	b := pp.buffer
-	dec := pp.decoder
+func (pp *prettyPrinter) formatObject(depth int) error {
+	if pp.iter.WhatIsNext() != jsoniter.ObjectValue {
+		return errors.New("next value is not an object")
+	}
+
+	pp.buffer.WriteString("{\n")
 
+	i := 0
 	for {
-		t, err := dec.Token()
-		if err == io.EOF {
-			return nil
+		s := pp.iter.ReadObject()
+		if s == "" {
+			break
 		}
-		if err != nil {
-			return fmt.Errorf("message Encode: unexpected error from Token(): %w", err)
+
+		if depth == 1 {
+			pp.topLevelFields = append(pp.topLevelFields, s)
 		}
-		switch v := t.(type) {
-
-		case json.Delim: // [ ] { }
-			switch v {
-			case ']':
-				fmt.Fprint(b, strings.Repeat("  ", depth-1))
-				fmt.Fprint(b, "]")
-				if dec.More() {
-					fmt.Fprint(b, ",")
-				}
-				fmt.Fprintf(b, "\n")
-				return nil
-			case '{':
-				fmt.Fprint(b, strings.Repeat("  ", depth))
-				fmt.Fprint(b, "{\n")
-				if err := pp.formatObject(depth + 1); err != nil {
-					return fmt.Errorf("formatArray(%d): decend failed: %w", depth, err)
-				}
-			case '[':
-				fmt.Fprint(b, strings.Repeat("  ", depth))
-				fmt.Fprint(b, "[\n")
-				if err := pp.formatArray(depth + 1); err != nil {
-					return fmt.Errorf("formatArray(%d): decend failed: %w", depth, err)
-				}
-			default:
-				return fmt.Errorf("formatArray(%d): unexpected token: %v", depth, v)
+
+		if i > 0 {
+			pp.buffer.WriteString(",\n")
+		}
+
+		pp.buffer.WriteString(strings.Repeat("  ", depth))
+		pp.writeString(s)
```

Comment:
We could speed this up by avoiding calling all the functions escaping the strings here (e.g. `unicodeEscapeSome`) in the hopes that nobody decided to include any weird characters in the key names but I don't know if that is a good idea.